spec.md

```diff
@@ -1,94 +1,120 @@
+The gateway must identify apps that match both the DNA hash specified in the request and have an installed app ID
+matching the `coordinator-identifier`. It must then check if the app ID is in the list of allowed apps configured in 
+`AllowedAppIds`. If multiple matching apps are found, an error should be returned since a unique app cannot be 
+determined.
 
-If no app is found in the initial check but the cache was populated, a fresh request to `ListApps` is made to update the cache, and the search is performed again. If the app still isn't found, or if it's found but not in the allowed list, an appropriate error is returned.
+If no app was found in the initial check but the cache was repopulated, the search is performed again. If the app still 
+isn't found, or if it's found but not in the allowed list, an appropriate error must be returned.
 
-### Prepare zome call signing
+### Connect to Holochain to make app calls
 
-At this point, we appear to have received a valid request that identifies an app running on Holochain. We should now try to issue signing credentials for ourselves.
+The gateway uses its admin API connection to list app interfaces with the `ListAppInterfaces` request. It looks for an 
+app interface that permits calls from the gateway or any location, and from the target app or any.
 
-The HTTP gateway uses ClientAgentSigner to manage signing credentials for zome calls, instead of using `sodoken` directly. The gateway then uses its admin API connection to authorize these credentials for each cell via authorize_signing_credentials. The granted functions are set according to the value of HC_GW_ALLOWED_FNS_{app-id}, either as All or a specific list of functions.
+If a matching interface cannot be found, a new interface must be provisioned using the admin API request 
+`AttachAppInterface`, and the `hc-http-gw` as an origin and with no app id specified.
 
-### Connect to Holochain to make app calls
+Using either the discovered or created app interface, a app connection is established. This is done by issuing a 
+connection token from the admin API with `IssueAppAuthenticationToken`. This token is then used to open an app 
+connection for the selected app. 
 
-The gateway uses its admin API connection to list app interfaces with the `ListAppInterfaces` request.
+After establishing the connection, the gateway authorizes signing credentials for each cell in the app and stores them
+locally. 
 
-It looks for an app interface that permits calls from the gateway or any location, and from the target app or any.
+> NOTE: The HTTP gateway uses a `ClientAgentSigner` from the [Rust client](https://github.com/holochain/holochain-client-rust)
```

Comment:
This is an implementation note for this implementation, so trying to make it appear a bit separately